docs/strings.md

```diff
@@ -0,0 +1,529 @@
+	 * The return value is a slice containing *at least* the characters in
+	 * `range` and starting from `range.index`. The `range` parameter will be
+	 * modified to define the range actually returned. The callee should either
+	 * allocate a new slice and copy the data into it, copy data into the
+	 * provided slice, or return an internal slice. 
+	 *
+	 * If the caller first calls `fastest_unicode_encoding` and uses the result
+	 * as the generic parameter then it will normally receive an internal buffer
+	 * when calling this method on immutable strings.
+	 *
+	 * This is intended to be used to build fast iterators.
+	 */
+	copy_or_view_data[Enc](self: readonly & String,
+	                       range: Range & mut)
+	                       buffer: Slice[Enc.CodeUnitType] & mut)
+	     : Slice[Enc.CodeUnitType] & (mut | imm)
+	     where Enc: StringEncoding;
+
+	/******************************************************************************
+	 * Operations on mutable strings
+	 *****************************************************************************/
+
+	/**
+	 * Updates a specific character. 
+	 */
+	update(self: (mut | iso) & String, index: U64, c: Rune);
+
+	/**
+	 * Replace the specified range in a string with another string.
+	 */
+	update_range[String Str](self: (mut | iso) & String,
+	                         range: Range,
+	                         newString: Str & imm);
+}
+```
+
+String views in Verona are a family of concrete types that do not own their data.
+The `String` interface is sufficient to express both strings that do and do not own their data.
+We therefore probably do not initially need a custom `StringView` interface.
+The Verona region system makes tracking ownership of string views relatively simple:
+
+ - By default, all mutable strings that own their data are `iso`.
+ - Mutable string views are `mut`, in the same region as their owning string.
+ - Immutable strings are globally immutable (`imm`).
+ - Immutable string views are immutable.
+
+This makes it easy to ensure that the lifetime of a string view is bound to the lifetime of the underlying string.
+Once stack types are better defined, it should also be possible for string views to be stack types associated with the region of the string.
+
+From the perspective of reasoning about mutation, there is no difference between a mutable string that has mutable string views and a mutable string view: if one or more string views exist on a `mut` string then that string is subject to mutation via other objects.
+If a function is passed an `iso & String` then the region guarantees are sufficient to know that the only other string views that may allow modification are `mut` parameters that are explicitly constrained to be allowed to be within the same region as the isolated string.
+Any function that takes two or more `mut & String` parameters is responsible for ensuring that they do not share the same backing store.
```

Comment:
This refers to Matt's discussion about `when (A, B)` where `A` and `B` could be references to the same region, if we want the users to check or if we can make the type system guarantee. Whatever that ends up being, this should be updated.